### DISCUSSION TITLE
Control Channels - Contribution

### DISCUSSION THREAD

**Discussion by shubhambooking2** [google/meridian#249] [General]
i have a model where i have multiple channels as part of control. Is there a way to get the summary indicating the contribution(%) associated with each of the control channel. Currently i see that we get the overall baseline contribution, however understanding i wanted to understand how much of contribution is associated with true baseline (comprising of trend, seasonality, knots etc.) and seperately the contribution of each of other control variables

**Comment by brettackers** [comment]
+1 on this.
we have a control where we had a major site update that saw a large increase in our KPI, however, the contribution output doesn't reflect this

**Comment by gauravsab1** [comment]
The contribution of control variables can be computed using the `analyzer.expected_impact` [method](https://github.com/google/meridian/blob/main/meridian/analysis/analyzer.py#L2075). You can iteratively set each control to 0 (or any other reasonable baseline value) and compare to the case where that same control is set to its usual value in the dataset. 

Note that the "overall baseline" (the orange bar in [this](https://developers.google.com/meridian/docs/advanced-modeling/interpret-visualizations) chart) is computed as expected revenue in the absence of all media channels (all media set to 0; all other variables held at the provided values in the dataset).

**Reply by gauravsab1** [reply]
Note: `expected_impact` was renamed `expected_outcome` today.

**Reply by brettackers** [reply]
Thanks @gauravsab1 , the link to the chart that you shared appears to only be accessible by Google employees

**Reply by gauravsab1** [reply]
Here is the correct link: https://developers.google.com/meridian/docs/advanced-modeling/interpret-visualizations. 

Let me know if this still doesn't work.

**Reply by gmoosman** [reply]
It would be great if someone could share some sample code on the `analyzer.expected_impact` method.

**Reply by gauravsab1** [reply]
Does [this](https://developers.google.com/meridian/reference/api/meridian/analysis/analyzer/Analyzer?hl=en#expected_outcome) help?

**Reply by gmoosman** [reply]
It helps somewhat, thank you!

If anyone has written a neat wrapper to get the contributions of each control and, maybe eve, chart them together with contributions for each media, please share ;)
e.g. a "decomp chart" with layered contribution of each input.

**Reply by sonriks6** [reply]
Hello,

This is what I use so far:

outcome_tf = model_analysis.expected_outcome(
    aggregate_times=False,
    use_kpi=True   
)
outcome = pd.DataFrame(tf.reduce_mean(outcome_tf, axis=[0, 1]).numpy())
outcome.to_csv(f'./{brand}_{market}_incremental_kpi.csv')

But have no clue on how to get the %contribution into the baseline of the control factors I'm using, any suggestions??

**Reply by gmoosman** [reply]
Thanks, @sonriks6 !
That is useful.

I am not used to the process. Do you have any code on how to set specific controls to 0 in order to compare to the case where that same control is set to its usual value in the dataset? (as @gauravsab1 writes)

**Reply by Amanda-MCN** [reply]
Having read [this in the documentation](https://developers.google.com/meridian/docs/advanced-modeling/control-variables#rsns) I just want to check that this solution is valid

**Reply by dattatreyam23** [reply]
Hi @Amanda-MCN,

Thank you for reaching out to the Google Meridian support team.

The proposed solution, using the `analyzer.expected_outcome` method provides an approximate workaround for the estimation of baseline contributions of control variables by setting each control to zero. However, the theory  as mentioned in the documentation on [Baseline Breakdown]( https://developers.google.com/meridian/docs/advanced-modeling/control-variables#rsns ) still stands valid. Control variables in Meridian do not produce causal effects or baseline contribution percentages, as they are primarily utilised to improve prediction accuracy, rather than acting as additive components in expected outcome. Hence, although the workaround is useful for exploratory purposes, it should not be relied upon for precise analysis of baseline contribution.

Additionally, we have recently implemented the support for *organic media and non-media treatment variables*. The contribution breakdown now will show the contribution of this variable as well. Please check our [documentation](https://developers.google.com/meridian/docs/advanced-modeling/organic-and-non-media-variables) to get an in-depth view of the same.

Feel free to reach out if you have any further questions regarding this.

Thank you,

Google Meridian Support Team

**Reply by Amanda-MCN** [reply]
Thanks for the clarification. As another clarification, I assume I can use expected_outcome to get the knots contribution?

**Reply by cpulavarthi** [reply]
Hello @Amanda-MCN, 

Meridian uses a time-varying intercept approach to model time effects. The time effects from $T$ time periods are modelled with `knots` number of parameters where the number of knots is determined by user input. Lesser number of knots would imply using lesser number of parameters to capture the time effects which results in smoothening and dimensionality reduction while capturing these effects. The value of the number of knots should be in the range of `1 <= knots <= n_times`. If you choose a high value for the `knots` argument, each time period will get its own parameter and the effect of a given time period is estimated using only data from that time period. Naturally, there is a bias-variance trade–off associated with knot selection. Please check our documentation on [Setting Knots](https://developers.google.com/meridian/docs/advanced-modeling/setting-knots) for more detailed information regarding what `knots` are, how to choose an appropriate value, and how to set them in Meridian.

Knots are only used to capture the time effects in Meridian and don’t have any contribution towards the target KPI. Hence you cannot use the `expected_outcome` method to determine their contribution. The `expected_outcome` method calculates the expected value of the outcome (i.e., the target KPI) - `E(Outcome|Media, RF, Organic media, Organic RF, Non-media treatments, Controls)`. You may check our [API documentation](https://developers.google.com/meridian/reference/api/meridian/analysis/analyzer/Analyzer#expected_outcome) of this method for more information. The table available on this [page](https://developers.google.com/meridian/docs/advanced-modeling/organic-and-non-media-variables#differences_between_the_types_of_input_variables) shows you the different types of variables that have an effect (contribution) on the target KPI, namely, media, non-media and organic-media variables. 

Feel free to reach out if you have any questions.

Thank you

Google Meridian Support Team

**Comment by sonriks6** [comment]
Hello,

There's also the option to compute the gamma_c to get the control variable contribution:

gamma_C = meridian.inference_data.posterior.gamma_c
mean_chains = gamma_C.mean(dim=('chain', 'draw'))
print(f"{mean_chains}")

You get and array with those insights but not sure if this is percentage attributed or how to read the results(?)
